.github/workflows/scripts/add-codeowners-to-pr.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+#
+#   Copyright The OpenTelemetry Authors.
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# Adds code owners without write access as reviewers on a PR. Note that
+# the code owners must still be a member of the `open-telemetry`
+# organization.
+#
+# Note that since this script is considered a requirement for PRs,
+# it should never fail.
+
+set -euo pipefail
+
+if [[ -z "${PR:-}" ]]; then
+    echo "PR has not been set, please ensure it is set."
+    exit 0
+fi
+
+main () {
+    CUR_DIRECTORY=$(dirname "$0")
+    FILES=$(gh pr view "${PR}" --json files | jq -r '.files[].path')
+    COMPONENTS=$(grep -oE '^[a-z]+/[a-z/]+ ' < .github/CODEOWNERS)
+    REVIEWERS=""
+    LABELS=""
+    declare -A PROCESSED_COMPONENTS
+
+    for COMPONENT in ${COMPONENTS}; do
+        for FILE in ${FILES}; do
+            MATCH=$(echo "${FILE}" | grep -E "^${COMPONENT}" || true)
+
+            if [[ -z "${MATCH}" ]]; then
+                continue
+            fi
+
+            # If we match a file with a component we don't need to process the file again.
+            FILES=$(printf "${FILES}" | grep -v "${FILE}")
+
+            if [[ -v PROCESSED_COMPONENTS["${COMPONENT}"] ]]; then
+                continue
+            fi
+
+            PROCESSED_COMPONENTS["${COMPONENT}"]=true
+
+            OWNERS=$(COMPONENT="${COMPONENT}" bash "${CUR_DIRECTORY}/get-codeowners.sh")
+
+            if [[ -n "${OWNERS}" ]]; then
+                if [[ -n "${REVIEWERS}" ]]; then
+                    REVIEWERS+=","
+                fi
+                REVIEWERS+="$(echo "${OWNERS}" | sed 's/@//g' | sed 's/ /,/g')"
+            fi
+
+            # Convert the CODEOWNERS entry to a label
+            COMPONENT_CLEAN=$(echo "${COMPONENT}" | sed -E 's%/$%%')
+            TYPE=$(echo "${COMPONENT_CLEAN}" | cut -f1 -d '/' )
+            NAME=$(echo "${COMPONENT_CLEAN}" | cut -f2- -d '/' | sed -E "s%${TYPE}\$%%")
+
+            if [[ -n "${LABELS}" ]]; then
+                LABELS+=","
+            fi
+            LABELS+="${TYPE}/${NAME}"
+        done
+    done
+
+    gh pr edit "${PR}" --add-label "${LABELS}" || echo "Failed to add labels to #${PR}"
+
+    # Note that adding the labels above will not trigger any other workflows to
+    # add code owners, so we have to do it here.
+    gh pr edit "${PR}" --add-reviewer "${REVIEWERS}" || echo "Failed to add reviewers to #${PR}"
+}
+
+main || echo "Failed to run $0"
```

Comment:
I'm kinda tempted to allow this to fail if it can't assign anyone because it will show which components don't have valid code owners.

Reply:
I agree that's a plus, but we'd also need to be okay with it being a hard requirement that code owners be members of the OpenTelemetry organization since that's a requirement for being a reviewer. Membership is currently considered a prerequisite to being a code owner according to our CONTRIBUTING.md file, but I'm not sure how consistently it is applied.

Reply:
Its been applied consistently since add I believe, but I think there are previous situations that would fail this workflow

Reply:
@open-telemetry/collector-contrib-approvers @open-telemetry/collector-contrib-maintainers what do you think?

Reply:
I believe we should add another workflow validating CODEOWNERS file instead of raising an error here